src/EditorFeatures/Core/Implementation/PasteTracking/PasteTrackingService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.PasteTracking
+{
+    [Export(typeof(IPasteTrackingService)), Shared]
+    [Export(typeof(PasteTrackingService))]
+    internal class PasteTrackingService : IPasteTrackingService
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly ILightBulbBroker2 _lightBulbBroker;
+        private readonly ISuggestedActionCategorySet _refactoringsCategorySet;
+
+        [ImportingConstructor]
+        public PasteTrackingService(IThreadingContext threadingContext, ILightBulbBroker2 lightBulbBroker, ISuggestedActionCategoryRegistryService categoryRegistryService)
+        {
+            _threadingContext = threadingContext;
+            _lightBulbBroker = lightBulbBroker;
+            _refactoringsCategorySet = categoryRegistryService.AllRefactorings;
+        }
+
+        public bool TryGetPastedTextSpan(Document document, out TextSpan textSpan)
+        {
+            if (!TryGetTextBuffer(document, out var textBuffer))
+            {
+                textSpan = default;
+                return false;
+            }
+
+            // `PropertiesCollection` is thread-safe
+            return textBuffer.Properties.TryGetProperty(this, out textSpan);
+        }
+
+        internal void RegisterPastedTextSpan(ITextView textView, ITextBuffer textBuffer, TextSpan textSpan)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            textBuffer.Changed += RemovePastedTestSpan;
+            textBuffer.Properties.AddProperty(this, textSpan);
```

Comment:
we have a userful type called PerSubjectBufferProperty that you probably want to use here.

Reply:
it will handle all the goodness about registering/unregistering things.

Reply:
When using those related extension methods the pasted text span gets cleared from properties when the TextView that added the property is closed. This causes a few of my tests which use linked files to fail since they were still expecting to have the pasted span.

Reply:
> When using those related extension methods the pasted text span gets cleared from properties when the TextView that added the property is closed

I can live with that :)  The expected use case for me is:

1. user pastes
2. user triggers lightbulb

If the user does:

1. user pastes
2. user closes view
3. user tries to trigger lightbulb

I'm ok with that now working :)

Reply:
Because the PerSubjectBufferProperty is exposed as extension methods for ITextView, I either need to change the interface for the service and move it from Features to EditorFeatures or I have to try to find an open TextView for the passed in Document. Is one option more preferable to the other?

Reply:
Gotcha.  Ok, let's clean thi sup a bit.   Let's ignore all of the above.  Instead, we'll do something similar to what you have now, except: 

1. don't use 'this' as the key to the property bag.  no need to potentially keep things alive or make references between this stuff.
2. i'm worried about double-pasting.  what happens in that case?  do we register the same callback twice?  when the callback fires, does it remove one of the registered event handlers?

Reply:
1. I am not sure I am seeing the issue you are with using the PasteTrackingService as the key. Could you tell me how it is a bad choice and maybe suggest an alternative?

2. Upon pasting for the second time, TextBuffer.Changed is invoked and the event handler is disconnected before a new TextSpan is registered.

Reply:
> I am not sure I am seeing the issue you are with using the PasteTrackingService as the key. 

It mostly feels skeezy.  Having our internal objects be keys in this property bag leaks us out in ways i'm not too comfortable with.  But it's not a direct problem per-se :)

> Upon pasting for the second time, TextBuffer.Changed is invoked and the event handler is disconnected before a new TextSpan is registered.

Ah ok.  Good to know.  Do we have tests for that?  Thanks!

Reply:
1. Would using `typeof(PasteTrackingService)` instead of `this` be a better key choice?
2. You might have to teach me a thing. =) What would a test for that look like? I do have a test for multiple pastes consecutively, but it only validates the tracked span matches the last paste operation. I'm not sure how I would track whether the event handler gets double registered.